feat(track): match on track duration

ali-ramadhan opened the pull request:
- [x] I have read the contributing guide in the documentation.

### Description

Added a `duration` property to the `MetaTrack` class. `get_match_value` will now consider track duration using a tolerance-based penalty system where a duration mismatch of 2.5% or less is not penalized, and the penalty increases linearly from 0 at 2.5% to 1 at 10% mismatch. Let me know what you think of this heuristic.

Also updated the tests. I had to pass `exists=True` to a lot of track and album factories (and enable the write plugin in some tests) as the import process and track matching now calls `Track.duration` which expects a file to exist from which to pull out a duration. There was only one test where I had to resort to patching.

Once this PR is merged, I'll open a PR for moe_musicbrainz to populate `MetaTrack.duration` and we should be up and running!

This PR is a result of discussions had in #313.

<!-- readthedocs-preview mrmoe start -->
----
📚 Documentation preview 📚: https://mrmoe--317.org.readthedocs.build/en/317/

<!-- readthedocs-preview mrmoe end -->

ali-ramadhan commented:
Sorry for the force pushes! I rebased off main with your latest refactor and resolved conflicts. Then I realized I needed to add a column for track duration in the import CLI. I probably need to add some tests so I've marked this PR as a draft, but would appreciate any thoughts/feedback you have @jtpavlock!

jtpavlock reviewed (changes requested):
No worries about the changes, thanks for leaving a comment/marking as a draft. I started a review so I have a few comments already, but I'll finish whenever you're ready. Besides the comments I've made, I'd also like to see a few tests for the new track duration testing logic through `get_match_value`.

ali-ramadhan commented on moe/util/core/match.py:
```diff
@@ -34,10 +34,47 @@
+BOTH_MISSING_DATA_PENALTY = 0.1
+ONE_MISSING_DATA_PENALTY = 0.2
+DURATION_TOLERANCE_THRESHOLD = 0.025
+DURATION_MAX_PENALTY_THRESHOLD = 0.10
```
Good point. Done!

ali-ramadhan commented on moe/util/core/match.py:
```diff
@@ -66,11 +103,17 @@ def get_match_value(
+        elif (
+            field == "duration"
+            and isinstance(value_a, (int, float))
+            and isinstance(value_b, (int, float))
```
That makes sense. Done! Agree that Moe shouldn't worry about all the possible things plugins can do.

ali-ramadhan commented:
Thanks for the quick review as usual @jtpavlock! I've addressed your comments so far and added a few tests.

You'll notice the new tests do things a bit differently: since we can't set/change durations on `Track`s such as ones from `track_factory`, I directly create some `MetaTrack`s and set their duration.

jtpavlock commented on moe/library/track.py:
```diff
@@ -116,6 +116,7 @@ def read_custom_tags(
     track_fields["composer"] = audio_file.composer
     track_fields["composer_sort"] = audio_file.composer_sort
     track_fields["disc"] = audio_file.disc
+    track_fields["duration"] = audio_file.length
```
We shouldn't be setting the duration field here, unless I'm missing something.

jtpavlock commented on moe/util/core/match.py:
```diff
@@ -34,10 +34,51 @@
+    min_penalty_threshold = 0.025
+    max_penalty_threshold = 0.10
+
+    if duration_a <= 0.0 and duration_b <= 0.0:
```
It doesn't seem like this check or the ONE_MISSING_DATA_PENALTY code is actually reachable because 0.0 is a False value so it's covered in `get_match_value`. You should be able to remove both conditions and then make the missing_data_penalty variables local to `get_match_value`.

jtpavlock commented on tests/util/core/test_match.py:
```diff
@@ -156,3 +157,60 @@ def test_diff_track(self):
+        track2 = MetaTrack(album=album, title="Test Track", track_num=1, disc=1)
+
+        track1.duration = 180.0
+        track2.duration = 184.0
```
This is a good test, but I think it would be more easily understandable if track 2's durations were set to a percentage of track 1's duration.That way you could see where you are on the scale instead of having to try and think if 184 is within 2.5% of 180.  That plus a few comments explaining the logic again such as "track 2's duration is > 10% from track 1 therefore the full match penalty is applied" or something.

jtpavlock reviewed (changes requested):
Nice work, a few minor things left I think and it should be good to go. Could you also include a screenshot of the import cli output with a few of the cases? Mainly I think:
1. track durations match
2. track durations don't match
3. one or both track durations missing

jtpavlock reviewed (commented):
I have no code changes left to suggest, this is in good shape. I think after reviewing the screenshots though, I'd like to discuss the UI changes a bit. 

I think right now displaying the changes as you've done might be a misleading as we aren't actually changing any track durations as is implied by the "->". Durations are inherent to the file itself and aren't metadata, all we're doing is comparing them. 

With that said, is it useful to even display the durations? They're only used for matching logic, but not actually applied changes like the rest of the fields I guess is what I'm getting at. Are you finding them useful for ensuring you have the right album to match?

ali-ramadhan commented:
> I think right now displaying the changes as you've done might be a misleading as we aren't actually changing any track durations as is implied by the "->". Durations are inherent to the file itself and aren't metadata, all we're doing is comparing them. 

This is a good point, and I remember it confusing me in beets as well. Maybe I'm used to the meaning now haha.

I'm trying to think if there's a clearer way of showing this information. Maybe just color the duration in green if it's a perfect match? For imperfect matches maybe print the file duration in white with the candidate track's duration in parentheses and colored yellow or red based on how bad the match is?

> With that said, is it useful to even display the durations? They're only used for matching logic, but not actually applied changes like the rest of the fields I guess is what I'm getting at. Are you finding them useful for ensuring you have the right album to match?

I personally do find them useful! I do like to check the mismatch and if it's a few seconds then I don't worry about it. But if the mismatch is large, maybe 30 seconds but  sometimes 2-3 minutes, then I check to see if my file is incomplete or corrupt or if I'm matching with the wrong release.

But I am probably biased since I always want more information haha.

If we feel like the duration column is creating clutter then would adding an option to include/exclude it be a good approach?

jtpavlock commented:
Okay, I like the idea of keeping them, but changing the formatting a bit to better illustrate that the duration isn't actually changing. I'm fine with keeping the duration column by default, and later down the line if it comes up we can add a config option to hide it, but I don't think that's necessary here at least.

For the formatting, I think going away from the "->" notation is the right call. Perhaps if it's a perfect match (or maybe even within 2.5% tolerance) just display a single duration in green (probably the file's makes sense).  What do you think of `5:32 (src: 5:15)` or `5:32 (ext: 5:14)` to explicitly state the source or the external source. Other options include just `5:32 (5:14)` or `5:32 | 5:14`, but not sure if those are clear enough about which duration is the file and which is imported.

The color coding for a perfect to bad match of green -> yellow -> red seems like a good idea. If there's no match data i.e. the import source doesn't come with a duration, perhaps just keep it white?

Let me know what you think makes the most sense.

ali-ramadhan commented:
@jtpavlock Thanks for the suggestion! Just rebased on top of `main` and I went with naming the column `Duration (external)` as I think `external` is clearer than `source`. Then I colored the external duration based on the match penalty. Here are a couple of screenshots showing what the new interface looks like.

<img width="1958" height="776" alt="Screenshot_20250716_212354_cropped" src="https://github.com/user-attachments/assets/847f6253-cc99-4b83-8eb3-0bfab897f8dc" />
<img width="1919" height="398" alt="Screenshot_20250716_213357" src="https://github.com/user-attachments/assets/6cacc122-19d4-4a29-9a89-5307b8c0231f" />

jtpavlock reviewed (changes requested):
The new formatting I think is much more easily readable. The color coding is nice too! 

I'm sorry to keep requesting changes, but I just have two things left, I think:
1. Now that `_duration_penalty` is used outside the match module, I don't think it makes sense to keep it a private function. I actually think it makes more sense to pull the field matching logic into its own new public function called `get_field_match_value` or something returning the match value between two _fields_. I think this UI change already shows that exposing this logic would be useful, and I'm sure it could be useful in other ways besides just this track match color coding too. This new api function should return the unweighted penalty/match value between two fields, and because this public function would now be part of the api, it should be tested directly.

I'm open to discussing the signature for this new function, but maybe something like:
```python
def get_field_match_value(value_a: Any, value_b: Any, field: FieldType)
```
Where `FieldType` would be an enum to specify the matching logic used such as `str`, `int`, `duration`, etc.

2. Can you add a test to `TestImportCLIOutput` in `test_import_cli` that displays the difference cases of duration matching, i.e. no match, perfect match, etc.? 

Edited to change recommendation to a generic api function

ali-ramadhan commented:
Thanks for the feedback @jtpavlock and apologies for the slow response!

I took a stab at adding a `FieldType` enum and a public `get_field_match_value` for string, int, and duration values. I also added some tests. Re-requesting a review to see what you think.

ali-ramadhan commented:
I got ruff to pass but not sure how to address this pyright complaint

```
  /home/alir/code/Moe/moe/util/core/match.py:87:40 - error: Argument of type "float | str | None" cannot be assigned to parameter "x" of type "ConvertibleToFloat" in function "__new__"
    Type "float | str | None" is not assignable to type "ConvertibleToFloat"
      Type "None" is not assignable to type "ConvertibleToFloat"
        "None" is not assignable to "str"
        "None" is incompatible with protocol "Buffer"
          "__buffer__" is not present
        "None" is incompatible with protocol "SupportsFloat"
          "__float__" is not present
        "None" is incompatible with protocol "SupportsIndex"
    ... (reportArgumentType)
```

without making this change:

```diff
diff --git a/moe/util/core/match.py b/moe/util/core/match.py
index 8ebe616..9f0f9dd 100644
--- a/moe/util/core/match.py
+++ b/moe/util/core/match.py
@@ -84,6 +84,8 @@ def get_field_match_value(
     if field_type == FieldType.STRING:
         return 1 - difflib.SequenceMatcher(None, str(value_a), str(value_b)).ratio()
     if field_type == FieldType.DURATION:
+        if value_a is None or value_b is None:
+            return one_missing_data_penalty
         return _duration_penalty(float(value_a), float(value_b))
     if field_type == FieldType.INTEGER:
         return 0 if value_a == value_b else 1
```

but it's redundant as it's already checked for a few lines above.

jtpavlock reviewed (changes requested):
Thanks for coming back to this! This PR certainly ended up expanding a bit beyond what was I think initially envisioned, so I appreciate you sticking around to see it through. A handful of comments mostly around I think simplifying the field_type logic a bit to fallback on using the same matching logic we were using before. 

We can carry-over the matching type logic from the original function where we simply inspected the type of value_a to determine the type of matching. I think the enum could be useful to specify non-default matching types like the duration type. With that, I think the only one we should hardcode for now is the duration field and everything else can be determined by just inspecting the value type like we were doing before. Hopefully that makes sense.

Also, could you please rebase off of main, I'd like to use the codecov fix to verify the test coverage here.

jtpavlock committed:
feat: importing album's will now display and consider track duration

When importing an album, each track's duration will be displayed in the import prompt and color-coded according to how well it matches the duration of the information being imported.

This feature also includes a new api function `get_field_match_penalty` which returns the penalty between two field's values.

ali-ramadhan reviewed (commented):
Thank you for pushing this PR to completion! I appreciate it and definitely looking forward to using this with v3.

I obviously can't approve or request changes on my own PR lol, but just pointing out a typo/bug. I feel like a simple test should have caught this though?

jtpavlock commented on moe/util/core/match.py:
```diff
@@ -34,11 +40,89 @@
+        slightly larger 0.2 penalty is returned.
+    """
+    both_missing_data_penalty = 0.1
+    one_missing_data_penalty = 0.2
```
Great catch! I actually also had the tests mixed up as well which is why it was not caught. I think it makes sense to have a higher penalty if just one value is missing because if both values are missing it's more of a match.
